authentication/src/main/java/io/camunda/authentication/filters/AdminUserCheckFilter.java

```diff
@@ -0,0 +1,93 @@
+import io.camunda.security.configuration.SecurityConfiguration;
+import io.camunda.service.RoleServices;
+import io.camunda.zeebe.protocol.record.value.DefaultRole;
+import io.camunda.zeebe.protocol.record.value.EntityType;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+public class AdminUserCheckFilter extends OncePerRequestFilter {
+
+  public static final String ADMIN_ROLE_ID = DefaultRole.ADMIN.getId();
+  public static final String USER_MEMBERS = "users";
+  private static final Logger LOG = LoggerFactory.getLogger(AdminUserCheckFilter.class);
+  private final SecurityConfiguration securityConfig;
+  private final RoleServices roleServices;
+
+  public AdminUserCheckFilter(
+      final SecurityConfiguration securityConfig, final RoleServices roleServices) {
+    this.securityConfig = securityConfig;
+    this.roleServices = roleServices;
+  }
+
+  @Override
+  protected void doFilterInternal(
+      final HttpServletRequest request,
+      final HttpServletResponse response,
+      final FilterChain filterChain)
+      throws ServletException, IOException {
+    if (request.getRequestURI().endsWith("/identity/setup")) {
+      filterChain.doFilter(request, response);
+      return;
+    }
+
+    final var hasConfiguredAdminUser =
+        !securityConfig
+            .getInitialization()
+            .getDefaultRoles()
+            .getOrDefault(ADMIN_ROLE_ID, Map.of())
+            .getOrDefault(USER_MEMBERS, Set.of())
+            .isEmpty();
+
+    if (hasConfiguredAdminUser) {
+      LOG.debug("Admin user has been configured.");
+      filterChain.doFilter(request, response);
+      return;
+    }
+
+    try {
+      final var adminRoleHasMembers =
+          roleServices
+                  .searchMembers(
+                      RoleQuery.of(
+                          builder ->
+                              builder.filter(
+                                  filter ->
+                                      filter
+                                          .joinParentId(ADMIN_ROLE_ID)
+                                          .memberType(EntityType.USER))))
+                  .total()
+              > 0;
+
+      if (!adminRoleHasMembers) {
+        LOG.debug("No user with admin role exists. Redirecting to identity setup page.");
+        response.sendRedirect(String.format("%s/identity/setup", request.getContextPath()));
+        return;
+      }
+    } catch (final RuntimeException ex) {
+      // Don't redirect the request if an error occurs while checking for admin role members.
+      LOG.error(
+          "Error while searching for admin role members. This might indicate that secondary storage is down.",
```

Comment:
Note: this will be expected in the future! there is a project currently starting off that allows cluster deployments without secondary storage :) could we make this more resilient already?

Reply:
I don't like downgrading the log. If there's no secondary storage they shouldn't do the query here at all. I'll leave it to that project team to resolve that here 😄